[filters] Add RBAC condition fuzzer (CEL expressions)

asraa opened the pull request:
This adds a fuzzer for CEL expression matching. These conditions are used in the RBAC filter for complementing the existing principal/permission model.

About a quarter of the execution time is coming from `google::api::expr::runtime::RegisterBuiltinFunctions`. The test runs locally at about 250 exec/sec.

See: https://github.com/envoyproxy/envoy/pull/7716

Risk Level: Low
Testing: Converted unit tests into corpus entries

asraa commented:
/review @kyessenov

kyessenov commented on source/extensions/filters/common/expr/context.cc:
```diff
@@ -157,9 +157,13 @@ absl::optional<CelValue> PeerWrapper::operator[](CelValue key) const {
+      }
     } else {
-      return CelValue::CreateString(info_.downstreamRemoteAddress()->asStringView());
+      if (info_.downstreamLocalAddress() != nullptr) {
```
downstreamRemoteAddress?

kyessenov commented on test/extensions/filters/common/expr/evaluator_fuzz_test.cc:
```diff
@@ -0,0 +1,72 @@
+
+    // Create the CEL expression.
+    Protobuf::Arena arena;
+    Expr::BuilderPtr builder = Expr::createBuilder(&arena);
```
You can move the builder out of the loop. It reuses the constant arena, so you might want to disable constant folding to reduce interference.

kyessenov commented on source/extensions/filters/common/expr/context.cc:
```diff
@@ -157,9 +157,13 @@ absl::optional<CelValue> PeerWrapper::operator[](CelValue key) const {
     if (local_) {
-      return CelValue::CreateString(info_.downstreamLocalAddress()->asStringView());
+      if (info_.downstreamLocalAddress() != nullptr) {
+        return CelValue::CreateString(info_.downstreamLocalAddress()->asStringView());
```
This is fine, but the stream info header states both can never be null.

asraa replied:
Ah OK. Should be fine then now since `TestStreamInfo` sets both. Originally it caught a ubsan bug when it wasn't set. Changed, since the fuzzer shouldn't hit that case.

kyessenov commented on source/extensions/filters/common/expr/context.cc:
```diff
@@ -96,7 +96,7 @@ absl::optional<CelValue> ResponseWrapper::operator[](CelValue key) const {
     }
   } else if (value == Size) {
-    return CelValue::CreateInt64(info_.bytesSent());
+    return CelValue::CreateUint64(info_.bytesSent());
```
Can we leave this to be int64? It makes inter-op with other int64 values better. Otherwise, we have to cast in the expressions to int64, which is annoying.

asraa commented:
@kyessenov are comprehension expressions supported? It seems like they aren't because `ListKeys` isn't implemented for e.g. with request maps.
Is it documented which expressions aren't supported? I might throw an `EnvoyException` in `createExpression` if they aren't supported.

kyessenov commented:
@asraa Comprehensions are explicitly disabled as of https://github.com/envoyproxy/envoy/pull/8746 since they can cause non-linear resource complexity. Same for string/list concatenations. I think you should get an exception building an expression for comprehensions, but not string/list concatenations. That's because string/list concat shares the operator + with normal addition (which is a language issue, that is being discussed). In any case, the evaluator should still successfully produce an error as a result of create expression or eval expression.

asraa commented:
Gotcha thanks! Thanks for that PR. I updated the branch and added the testcase that was failing from the comprehension expressions. Should be good now.

kyessenov commented:
Changes look fine to me, but I'm not very familiar with the fuzzing framework. Let me know if the fuzzer detects something.

asraa commented:
Will do. After about a day of being checked in to upstream, OSS-Fuzz will report any crashes that come up. I'll reach out to you with anything that comes up. Also feel free to ping me if you want to extend fuzzers for the filter.

kyessenov commented on source/extensions/filters/common/expr/context.cc:
```diff
@@ -187,6 +187,7 @@ absl::optional<CelValue> PeerWrapper::operator[](CelValue key) const {
   if (value == Address) {
     if (local_) {
       return CelValue::CreateString(info_.downstreamLocalAddress()->asStringView());
+
```
extra new line

htuch commented on test/extensions/filters/common/expr/evaluator_fuzz_test.cc:
```diff
@@ -0,0 +1,74 @@
+
+DEFINE_PROTO_FUZZER(const test::extensions::filters::common::expr::EvaluatorTestCase& input) {
+  // Create builder without constant folding.
+  static Expr::BuilderPtr builder = Expr::createBuilder(nullptr);
```
Is this stateless? I.e. it won't vary between test cases and so is safe to make static? If so, please update the above comment.

kyessenov replied:
Builder can be shared between expressions.

asraa replied:
Updated comment. Yep, it was time-consuming to create the builder between iterations as well, so I felt it was a worthy object to make static.

htuch commented on test/extensions/filters/common/expr/evaluator_fuzz_test.cc:
```diff
@@ -0,0 +1,74 @@
+    Http::TestHeaderMapImpl response_trailers = Fuzz::fromHeaders(input.trailers());
+
+    // Evaluate the CEL expression.
+    Protobuf::Arena arena;
```
What is this for?

kyessenov replied:
CEL creates all temporary objects on an arena. I think this is used for this purpose.

asraa replied:
Updated, passed in to `evaluate`. I was playing around with which would be faster.

asraa commented:
Sorry it passed through my attention until it was marked closed!! Should be ready now.
